[BUGFIX] Restore support for V2 API style custom expectation rendering

jdimatteo opened the pull request:
Restores at least partial support for custom expectation rendering, as it worked in Great Expectations version 0.12.7.

Fixes #3174.

Changes proposed in this pull request:
- If no V3 API style render is found and a V2 API style custom render is found, do translation to match the legacy rendering signature.
- If no V3 API style render is found and a V2 API style custom render is found, fall back to legacy way of displaying observed values.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

jdimatteo committed:
Restore support for V2 API style custom expectation rendering

FIXES:3174

Restores at least partial support for custom expectation rendering.

jdimatteo committed:
Add changelog entry

jdimatteo committed:
Run v2 api style rendering example as part of integration tests

jdimatteo committed:
cleanup

jdimatteo committed:
Cleanup / add comments

jdimatteo committed:
Merge branch 'develop' into BUGFIX/restore-custom-expectation-rendering-with-v2-api

jdimatteo commented:
There are [two snowflake test failures](https://dev.azure.com/great-expectations/great_expectations/_build/results?buildId=12911&view=logs&j=0be527af-e1cb-5224-1725-2baa5917c313&t=096d6768-99a4-5ec0-6a65-65122b15571c&l=1009) that are not caused by this PR

jdimatteo committed:
Merge branch 'develop' into BUGFIX/restore-custom-expectation-rendering-with-v2-api

talagluck commented on great_expectations/render/renderer/content_block/validation_results_table_content_block.py:
```diff
@@ -168,3 +179,48 @@ def row_generator_fn(
+        def expectation_string_fn_with_legacy_translation(
+            configuration: ExpectationConfiguration, runtime_configuration: dict
+        ):
+            # With the V2 API, the epectation string function had a different signature; below translates from the new signature to the legacy signature.
```
```suggestion
            # With the V2 API, the expectation string function had a different signature; the below translates from the new signature to the legacy signature.
```

talagluck commented on great_expectations/render/renderer/content_block/validation_results_table_content_block.py:
```diff
@@ -168,3 +179,48 @@ def row_generator_fn(
+        warnings.warn(
+            "V2 API style custom rendering is deprecated and is not fully supported anymore; please switch to V3 API and associated rendering style",
+            DeprecationWarning,
+        )
```
I just want to confirm - are the `_get_legacy_v2_api_observed_value` and  `_get_legacy_v2_api_style_expectation_string_fn` called together? I want to make sure we're not getting double DeprecationWarnings.

jdimatteo replied:
Good point -- I added a commit to fix this.

Yes, they are called together. I removed DeprecationWarning from `_get_legacy_v2_api_observed_value`, since it is only called if `_get_legacy_v2_api_style_expectation_string_fn` is executed, which already gives a DeprecationWarning.

talagluck reviewed (commented):
Hi @jdimatteo - thank you, this looks awesome! I love how thorough you were with the tests. Could you explain why you decided to add it as a Docusaurus integration test? Typically, we only add Docusaurus integration tests like this when there are corresponding docs which use the test. In those cases, the test files are more like flattened scripts, with some comments, and easily followable by a user. 

I think this test might make more sense as one or more unit tests within the `tests/render` directory. Do you have strong feeling about this?

jdimatteo committed:
Remove duplicate DeprecationWarning

Removed DeprecationWarning from _get_legacy_v2_api_observed_value, since it is only called if _get_legacy_v2_api_style_expectation_string_fn is executed, which already gives a DeprecationWarning

jdimatteo committed:
Merge remote-tracking branch 'upstream/develop' into BUGFIX/restore-custom-expectation-rendering-with-v2-api

jdimatteo committed:
Fix typo with @talagluck's suggested change

Co-authored-by: talagluck <tal@superconductive.com>

jdimatteo committed:
Add unit test

jdimatteo committed:
remove Docusaurus integration test

jdimatteo commented:
@talagluck thanks for the review and pointers on the tests!

> I think this test might make more sense as one or more unit tests within the `tests/render` directory.

Latest commit adds these tests and removes the Docusaurus integration tests.

Please take another look!

talagluck reviewed (approved):
Thanks for the updates incorporating all of the feedback, @jdimatteo - LGTM!

talagluck commented:
I'm waiting for another PR to merge, and then I'll update and merge this.

talagluck committed:
Merge branch 'develop' into BUGFIX/restore-custom-expectation-rendering-with-v2-api
